### PR TITLE
allow addons to control which recipes are registered

### PR DESCRIPTION
**Pull request by TechLord22**
## What
This PR allows addons to control which recipes GT or other addons register. Each major group of recipes now has an event fired for checking if it should be registered. Other mods can choose to stop GT from registering the recipes accordingly, and also implement this system for themselves, allowing other mods to cancel addons' recipes as well.

This is potentially useful for modpacks with core mods, which want to disable a lot of GT's default content without manually removing everything.

## Outcome
Allows addons to control which recipes are registered.

### CODE REVIEW

File: src/main/java/gregtech/loaders/recipe/MachineRecipeLoader.java [hunk 1]
```diff
@@ -54,15 +54,15 @@ private MachineRecipeLoader() {
 
     public static void init() {
```
Comment (LAGIdiot) [inline src/main/java/gregtech/loaders/recipe/MachineRecipeLoader.java:55]: Registration is quite inconsistent and hard to follow. Like where whole init function should be registered and where just part of the file.

File: src/main/java/gregtech/loaders/recipe/GTRecipeLoaders.java [hunk 2]
```diff
@@ -0,0 +1,391 @@
+package gregtech.loaders.recipe;
+
+import gregtech.api.GTValues;
+import gregtech.api.recipes.loader.IRecipeLoader;
+
+import javax.annotation.Nonnull;
+
+public enum GTRecipeLoaders implements IRecipeLoader {
```
Comment (LAGIdiot) [inline src/main/java/gregtech/loaders/recipe/GTRecipeLoaders.java:8]: I feel this either needs to be grouped by virtual categories or in alphabetic order. As right now for someone who is it not using everyday it is hard to find right category in all of these.

**Review by LAGIdiot — COMMENTED** [review]
I like idea behind this PR. This will be definitely makes integration with GTCEu easier.

Will we know is someone cancels our registration? Will it be logged somewhere on global level? Because if not it will be hard to know what broke and why are recipes missing.

Is it possible to load GTCEu without any recipes? Does it run? As I feel there maybe possible crashes because of incorrect null/empty checks. Also those it really stars empty or there are some hardcoded which this does not cover?

**Review by LAGIdiot — CHANGES_REQUESTED** [review]
As per my previous comments.

**Comment by LAGIdiot** [comment]
There were some recipe changes (mostly moving them between files) in #1276 which may need to be reflected here

**Comment by serenibyss** [comment]
Should be no longer needed with #1847
